File: tools-excel/src/main/java/cn/gjing/tools/excel/valid/DefaultDateValidation.java
package cn.gjing.tools.excel.valid;


import org.apache.poi.ss.usermodel.DataValidation;
import org.apache.poi.ss.usermodel.DataValidationConstraint;
import org.apache.poi.ss.usermodel.DataValidationHelper;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.util.CellRangeAddressList;
import org.apache.poi.xssf.usermodel.XSSFSheet;

/**
 * Default time verifier
 *
 * @author Gjing
 **/
public class DefaultDateValidation implements ExcelDateValidation {
    @Override
    public void valid(ExcelDateValid excelDateValid, Sheet sheet, int firstRow, int lastRow, int firstCol, int lastCol) {
        DataValidationHelper helper = sheet.getDataValidationHelper();
        DataValidationConstraint dvConstraint;
        if (sheet instanceof XSSFSheet) {
            dvConstraint = helper.createDateConstraint(excelDateValid.operatorType().getType(), "date(" + excelDateValid.expr1().replaceAll("-", ",") + ")",
                    "date(" + excelDateValid.expr2().replaceAll("-", ",") + ")", excelDateValid.pattern());
        } else {
            dvConstraint = helper.createDateConstraint(excelDateValid.operatorType().getType(), excelDateValid.expr1(),
                    excelDateValid.expr2(), excelDateValid.pattern());
        }
        CellRangeAddressList regions = new CellRangeAddressList(firstRow, lastRow, firstCol, lastCol);
        DataValidation dataValidation = helper.createValidation(dvConstraint, regions);
        dataValidation.setShowErrorBox(excelDateValid.showErrorBox());
        dataValidation.setErrorStyle(excelDateValid.rank().getRank());
        dataValidation.createErrorBox(excelDateValid.errorTitle(), excelDateValid.errorContent());
        sheet.addValidationData(dataValidation);
    }
}

File: tools-excel/src/main/java/cn/gjing/tools/excel/valid/DefaultDateValidation.scala
package cn.gjing.tools.excel.valid

import org.apache.poi.ss.usermodel.Sheet
import org.apache.poi.ss.util.CellRangeAddressList

/**
 * Default time verifier
 * @author Gjing
 **/
class DefaultDateValidation extends ExcelValidation{
  /**
   * Custom time validation rules
   *
   * @param dateValid DateValid
   * @param sheet     sheet
   * @param firstRow  First row
   * @param firstCol  First col
   * @param lastCol   Last col
   */
  override def valid(dateValid: DateValid, sheet: Sheet, firstRow: Int, firstCol: Int, lastCol: Int): Unit = {
    val helper = sheet.getDataValidationHelper
    val dvConstraint = helper.createDateConstraint(dateValid.operatorType.getType, dateValid.expr1, dateValid.expr2, dateValid.pattern)
    val regions = new CellRangeAddressList(firstRow, if (dateValid.boxLastRow == 0) firstRow
    else dateValid.boxLastRow + firstRow, firstCol, lastCol)
    val dataValidation = helper.createValidation(dvConstraint, regions)
    dataValidation.setShowErrorBox(dateValid.showErrorBox)
    dataValidation.setShowPromptBox(dateValid.showPromptBox)
    dataValidation.setErrorStyle(dateValid.rank.getRank)
    dataValidation.createErrorBox(dateValid.errorTitle, dateValid.errorContent)
    sheet.addValidationData(dataValidation)
  }
}
